File: scheduler.js
class Employee {
    constructor(id, name, priority = 0) {
        this.id = id;
        this.name = name;
        this.priority = priority;
        this.vacationDays = [];
        this.preferredExclusionDays = [];
    }
}

class Shift {
    constructor(date, type, employeeId = null) {
        this.date = new Date(date);
        this.type = type;
        this.employeeId = employeeId;
        // Наряды 1-6 заканчиваются на следующий день, тип 7 - в тот же день
        this.endDate = type !== 7 ? 
            new Date(date.getTime() + 24 * 60 * 60 * 1000) : 
            new Date(date);
    }
}

class ShiftScheduler {
    constructor() {
        this.employees = [];
        this.assignedShifts = [];
        this.unassignedShifts = [];
        this.employeeStats = new Map();
    }

    addEmployee(name, vacationDays = [], preferredExclusionDays = [], priority = 0) {
        const employeeId = this.employees.length + 1;
        const employee = new Employee(employeeId, name, priority);
        
        employee.vacationDays = vacationDays.map(d => new Date(d));
        employee.preferredExclusionDays = preferredExclusionDays.map(d => new Date(d));
        
        this.employees.push(employee);
        this.employeeStats.set(employeeId, {
            shiftsCount: 0,
            occupiedDays: new Set(),
            monthlySlots: 15
        });
        
        return employeeId;
    }

    generateSchedule(dailyShifts) {
        this.assignedShifts = [];
        this.unassignedShifts = [];
        
        // Сбрасываем статистику
        this.employeeStats.clear();
        this.employees.forEach(emp => {
            this.employeeStats.set(emp.id, {
                shiftsCount: 0,
                occupiedDays: new Set(),
                monthlySlots: 15
            });
        });

        // Создаем все наряды
        const allShifts = [];
        Object.entries(dailyShifts).forEach(([dateStr, shiftTypes]) => {
            const date = this.parseDate(dateStr);
            shiftTypes.forEach(type => {
                allShifts.push(new Shift(date, type));
            });
        });

        // Сортируем по дате
        allShifts.sort((a, b) => a.date - b.date);

        // Распределяем наряды
        allShifts.forEach(shift => {
            const availableEmployees = this.getAvailableEmployees(shift);
            
            if (availableEmployees.length > 0) {
                // Выбираем лучшего сотрудника
                availableEmployees.sort((a, b) => {
                    const aStats = this.employeeStats.get(a.id);
                    const bStats = this.employeeStats.get(b.id);
                    
                    // Сначала по приоритету (высокий лучше)
                    if (a.priority !== b.priority) {
                        return b.priority - a.priority;
                    }
                    // Затем по количеству нарядов (меньше лучше)
                    return aStats.shiftsCount - bStats.shiftsCount;
                });

                const bestEmployee = availableEmployees[0];
                this.assignShiftToEmployee(shift, bestEmployee);
            } else {
                this.unassignedShifts.push(shift);
            }
        });
    }

    getAvailableEmployees(shift) {
        const shiftDays = this.getShiftDays(shift);
        const available = [];

        this.employees.forEach(employee => {
            const stats = this.employeeStats.get(employee.id);
            
            // 1. Проверка отпуска
            const isOnVacation = shiftDays.some(day => 
                employee.vacationDays.some(vacation => 
                    this.isSameDay(vacation, day)
                )
            );
            
            if (isOnVacation) return;

            // 2. Проверка лимита суточных нарядов
            if (shift.type !== 7 && stats.monthlySlots <= 0) return;

            // 3. Проверка занятости
            const isBusy = shiftDays.some(day => stats.occupiedDays.has(day.getTime()));
            if (isBusy) return;

            // 4. Проверка предпочтительных дней исключения (мягкое ограничение)
            const hasPreferredExclusion = shiftDays.some(day => 
                employee.preferredExclusionDays.some(excl => 
                    this.isSameDay(excl, day)
                )
            );

            available.push({
                employee,
                hasPreferredExclusion
            });
        });

        // Сначала возвращаем сотрудников без предпочтений исключения
        const withoutExclusion = available.filter(a => !a.hasPreferredExclusion).map(a => a.employee);
        if (withoutExclusion.length > 0) return withoutExclusion;

        // Если таких нет, возвращаем всех
        return available.map(a => a.employee);
    }

    assignShiftToEmployee(shift, employee) {
        shift.employeeId = employee.id;
        this.assignedShifts.push(shift);
        
        const stats = this.employeeStats.get(employee.id);
        stats.shiftsCount++;
        
        // Отмечаем занятые дни
        const shiftDays = this.getShiftDays(shift);
        shiftDays.forEach(day => {
            stats.occupiedDays.add(day.getTime());
        });
        
        // Уменьшаем слоты для суточных нарядов
        if (shift.type !== 7) {
            stats.monthlySlots--;
        }
    }

    getShiftDays(shift) {
        if (shift.type === 7) {
            return [shift.date];
        } else {
            return [shift.date, new Date(shift.date.getTime() + 24 * 60 * 60 * 1000)];
        }
    }

    isSameDay(date1, date2) {
        return date1.getDate() === date2.getDate() &&
               date1.getMonth() === date2.getMonth() &&
               date1.getFullYear() === date2.getFullYear();
    }

    parseDate(dateStr) {
        const [day, month, year] = dateStr.split('.').map(Number);
        return new Date(year, month - 1, day);
    }

    formatDate(date) {
        const day = String(date.getDate()).padStart(2, '0');
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const year = date.getFullYear();
        return `${day}.${month}.${year}`;
    }

    analyzeUnassignedReason(shift) {
        const shiftDays = this.getShiftDays(shift);
        
        // Проверяем сотрудников не в отпуске
        const availableEmployees = this.employees.filter(emp => 
            !shiftDays.some(day => 
                emp.vacationDays.some(vacation => this.isSameDay(vacation, day))
            )
        );

        if (availableEmployees.length === 0) return "Все сотрудники в отпуске";

        // Проверяем лимит
        if (shift.type !== 7) {
            const overloadedCount = availableEmployees.filter(emp => {
                const stats = this.employeeStats.get(emp.id);
                return stats.monthlySlots <= 0;
            }).length;

            if (overloadedCount === availableEmployees.length) return "Достигнут лимит нарядов";
        }

        // Проверяем занятость
        const busyCount = availableEmployees.filter(emp => {
            const stats = this.employeeStats.get(emp.id);
            return shiftDays.some(day => stats.occupiedDays.has(day.getTime()));
        }).length;

        if (busyCount === availableEmployees.length) return "Все сотрудники заняты";

        return "Недостаточно сотрудников";
    }

    getStatistics() {
        const totalRequested = this.assignedShifts.length + this.unassignedShifts.length;
        
        return {
            total_employees: this.employees.length,
            total_assigned: this.assignedShifts.length,
            total_unassigned: this.unassignedShifts.length,
            total_requested: totalRequested,
            employee_loads: this.employees.map(emp => {
                const stats = this.employeeStats.get(emp.id);
                return {
                    name: emp.name,
                    shifts: stats.shiftsCount,
                    remaining_slots: stats.monthlySlots
                };
            })
        };
    }
}

File: scheduler.py
from datetime import datetime, timedelta
from collections import defaultdict
from typing import List, Dict, Optional

class Employee:
    """Класс для представления сотрудника"""
    def __init__(self, id: int, name: str, priority: int = 0):
        self.id = id
        self.name = name
        self.priority = priority
        self.vacation_days: List[datetime] = []
        self.preferred_exclusion_days: List[datetime] = []

class Shift:
    """Класс для представления наряда"""
    def __init__(self, date: datetime, shift_type: int, employee_id: Optional[int] = None):
        self.date = date
        self.type = shift_type
        self.employee_id = employee_id
        # Наряды типа 1-6 заканчиваются на следующий день в 9:00
        # Наряд типа 7 (8 часов) заканчивается в тот же день
        self.end_date = date + timedelta(days=1) if shift_type in [1, 2, 3, 4, 5, 6] else date

class ShiftScheduler:
    """Основной класс для распределения нарядов"""
    def __init__(self):
        self.employees: List[Employee] = []
        self.assigned_shifts: List[Shift] = []
        self.unassigned_shifts: List[Shift] = []
        self.employee_stats = defaultdict(lambda: {
            'shifts_count': 0, 
            'occupied_days': set(), 
            'monthly_slots': 15
        })
    
    def add_employee(self, name: str, vacation_days: List[datetime] = None, 
                     preferred_exclusion_days: List[datetime] = None, priority: int = 0) -> int:
        """Добавить сотрудника в систему"""
        employee_id = len(self.employees) + 1
        employee = Employee(employee_id, name, priority)
        if vacation_days:
            employee.vacation_days = vacation_days
        if preferred_exclusion_days:
            employee.preferred_exclusion_days = preferred_exclusion_days
        self.employees.append(employee)
        return employee_id

    def generate_schedule(self, daily_shifts: Dict[datetime, List[int]]):
        """Распределение нарядов с учетом всех ограничений"""
        self.assigned_shifts = []
        self.unassigned_shifts = []
        
        # Сбрасываем статистику сотрудников
        for employee in self.employees:
            self.employee_stats[employee.id] = {
                'shifts_count': 0, 
                'occupied_days': set(), 
                'monthly_slots': 15
            }
        
        # Создаем список всех нарядов для распределения
        all_shifts = []
        for date, shift_types in daily_shifts.items():
            for shift_type in shift_types:
                all_shifts.append(Shift(date, shift_type))
        
        # Сортируем наряды по дате для последовательного распределения
        all_shifts.sort(key=lambda x: x.date)
        
        # Распределяем наряды
        for shift in all_shifts:
            available_employees = self._get_available_employees(shift)
            
            if available_employees:
                # Выбираем сотрудника с учетом приоритета и количества нарядов
                # Сотрудники с более высоким приоритетом получают наряды в первую очередь
                available_employees.sort(key=lambda emp: (
                    -emp.priority,
                    self.employee_stats[emp.id]['shifts_count']
                ))
                employee = available_employees[0]
                
                # Назначаем наряд
                shift.employee_id = employee.id
                self.assigned_shifts.append(shift)
                
                # Обновляем статистику
                self.employee_stats[employee.id]['shifts_count'] += 1
                
                # Отмечаем занятые дни
                shift_days = self._get_shift_days(shift)
                self.employee_stats[employee.id]['occupied_days'].update(shift_days)
                
                # Уменьшаем доступные слоты для суточных нарядов
                if shift.type in [1, 2, 3, 4, 5, 6]:
                    self.employee_stats[employee.id]['monthly_slots'] -= 1
            else:
                # Нет доступных сотрудников - добавляем в нераспределенные
                self.unassigned_shifts.append(shift)
    
    def _get_shift_days(self, shift: Shift) -> List[datetime]:
        """Возвращает список дней, которые занимает наряд"""
        if shift.type in [1, 2, 3, 4, 5, 6]:
            # Суточный наряд занимает текущий день и следующий
            return [shift.date, shift.date + timedelta(days=1)]
        else:
            # 8-часовой наряд занимает только текущий день
            return [shift.date]
    
    def _get_available_employees(self, shift: Shift) -> List[Employee]:
        """Поиск сотрудников, доступных для этого наряда"""
        available = []
        shift_days = self._get_shift_days(shift)
        
        for employee in self.employees:
            # 1. Проверка отпуска
            if any(any(vacation.date() == day.date() for vacation in employee.vacation_days) 
                   for day in shift_days):
                continue
            
            # 2. Проверка предпочтительных дат исключения (учитываем, но не блокируем жестко)
            # Если есть другие доступные сотрудники, пропускаем этого
            has_preferred_exclusion = any(
                any(excl.date() == day.date() for excl in employee.preferred_exclusion_days) 
                for day in shift_days
            )
            
            # 3. Проверка лимита суточных нарядов (15 в месяц)
            if shift.type in [1, 2, 3, 4, 5, 6] and self.employee_stats[employee.id]['monthly_slots'] <= 0:
                continue
            
            # 4. Проверка что сотрудник не занят в эти дни другими нарядами
            if any(day in self.employee_stats[employee.id]['occupied_days'] for day in shift_days):
                continue
            
            # Добавляем в список, отмечая предпочтения
            available.append((employee, has_preferred_exclusion))
        
        # Сначала пытаемся выбрать тех, у кого нет предпочтений исключения
        preferred_available = [emp for emp, has_excl in available if not has_excl]
        if preferred_available:
            return preferred_available
        
        # Если таких нет, возвращаем всех доступных
        return [emp for emp, _ in available]

    def analyze_unassigned_reason(self, shift: Shift) -> str:
        """Анализирует причину нераспределенного наряда"""
        shift_days = self._get_shift_days(shift)
        
        # Проверяем, есть ли вообще сотрудники не в отпуске в эти дни
        available_employees = []
        for emp in self.employees:
            if not any(any(vac.date() == day.date() for vac in emp.vacation_days) for day in shift_days):
                available_employees.append(emp)
        
        if not available_employees:
            return "Все сотрудники в отпуске"
        
        # Проверяем лимит суточных нарядов
        if shift.type in [1, 2, 3, 4, 5, 6]:
            overloaded_count = 0
            for emp in available_employees:
                if self.employee_stats[emp.id]['monthly_slots'] <= 0:
                    overloaded_count += 1
            
            if overloaded_count == len(available_employees):
                return "Все доступные достигли лимита (15 нарядов)"
        
        # Проверяем занятость
        busy_count = 0
        for emp in available_employees:
            if any(day in self.employee_stats[emp.id]['occupied_days'] for day in shift_days):
                busy_count += 1
        
        if busy_count == len(available_employees):
            return "Все доступные заняты в эти дни"
        
        return "Недостаточно сотрудников"

    def get_statistics(self) -> Dict:
        """Получить статистику распределения"""
        stats = {
            'total_employees': len(self.employees),
            'total_assigned': len(self.assigned_shifts),
            'total_unassigned': len(self.unassigned_shifts),
            'employee_loads': []
        }
        
        for employee in self.employees:
            employee_shifts = [s for s in self.assigned_shifts if s.employee_id == employee.id]
            stats['employee_loads'].append({
                'name': employee.name,
                'shifts': len(employee_shifts),
                'remaining_slots': self.employee_stats[employee.id]['monthly_slots']
            })
        
        return stats
